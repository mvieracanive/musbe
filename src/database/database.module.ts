import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { ConnectionOptions } from 'typeorm';

@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      //inject: [ConfigService],
      async useFactory() {
        
        return {
          type: 'sqlite',
          dropSchema: true,
          logging: false,
          database: 'musla',
          entities: [
            __dirname + '/../**/*.entity{.ts,.js}',],
          synchronize: true, //Delete on production
        } as ConnectionOptions;
      },
    }),
  ],
  exports: [TypeOrmModule],
})
export class DatabaseModule {}
