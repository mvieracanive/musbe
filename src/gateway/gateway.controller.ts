import { Body, Controller, Get, Param, Post, Query, Res } from "@nestjs/common";
import { plainToClass } from "class-transformer";
import { GatewayDto } from "../dtos/gateway.dto";
import { GatewayService } from "./gateway.service";

@Controller('gateway')
export class GatewayController{
    constructor (private _service: GatewayService){}
    @Post()
    async create(@Body() dto: GatewayDto, @Res({ passthrough: true }) res: any){
        const obj:GatewayDto= plainToClass(GatewayDto, {
                ...dto
                }, { enableImplicitConversion: true })

        const ret = await this._service.create(obj);
        res.location(`/gateway/${ret.serial}`)
        return ret;
    }

    @Get()
    getCollection(
        @Query('sort') sortby: string
    ): Promise<GatewayDto[]> {
        return this._service.getCollection(sortby);
    }

    @Get('/:id')
    getOne(
        @Param('id') id: string
    ): Promise<GatewayDto> {
        return this._service.getOne(id);
    }

}