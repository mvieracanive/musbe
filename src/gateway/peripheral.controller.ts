import { Body, Controller, Delete, Param, Post, Res } from "@nestjs/common";
import { PeripheralCreateDto } from "src/dtos/peripheral-create.dto";
import { PeripheralDto } from "src/dtos/peripheral.dto";
import { PeripheralService } from "./peripheral.service";

@Controller('gateway/:gatewayid/peripheral')
export class PeripheralController{
    constructor (private _service: PeripheralService){}
    @Post()
    async create(
        @Param('gatewayid') gateway: string, 
        @Body() dto:PeripheralCreateDto, 
        @Res({ passthrough: true }) res: any){
        const ret = await this._service.create(dto, gateway);
        res.location(`/gateway/${gateway}/pheriferal/${dto.uid}`)
        return ret;
    }

    @Delete('/:id')
    async delete(@Param('id') id: number){
        return this._service.delete(id);
    }
}