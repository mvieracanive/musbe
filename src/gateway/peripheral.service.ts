import { HttpException, HttpStatus, Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { PeripheralCreateDto } from "src/dtos/peripheral-create.dto";
import { PeripheralDto } from "src/dtos/peripheral.dto";
import { Repository } from "typeorm";
import { Peripheral } from "./entities/peripheral.entity";

Injectable()
export class PeripheralService{
    constructor(@InjectRepository(Peripheral) private _repo: Repository<Peripheral>){}
    
    async create(dto: PeripheralCreateDto, gateway: string):Promise<PeripheralDto>{
        let currenttotal = await this._repo.find({where: {gateway}});
        if (currenttotal.length === 10)
            throw new HttpException('The gatware has reached top number of peripherals', HttpStatus.CONFLICT)
        try{
            

            await this._repo.insert({...dto, gateway:{serial: gateway}});
        }
        catch(ex){
            console.log(ex);
            if (ex.errno == 19)
                throw new HttpException(
                'Peripheral uid muest be unique for the gateway',
                HttpStatus.CONFLICT,
                );
            throw new HttpException('Internal Error. Please check the logs', HttpStatus.INTERNAL_SERVER_ERROR)
        }
        return {
            ...dto,
            gateway
        };
    }

    async delete(id:number){

    }
}