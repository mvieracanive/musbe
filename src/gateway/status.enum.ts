enum PeripheralStatus{
    ONLINE = 'online',
    OFFLINE = 'offline' 
}