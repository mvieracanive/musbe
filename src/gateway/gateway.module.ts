import { TypeOrmModule } from "@nestjs/typeorm";
import { Module } from "@nestjs/common";
import { Gateway } from "./entities/gateway.entity";
import { Peripheral } from "./entities/peripheral.entity";
import { GatewayController } from "./gateway.controller";
import { GatewayService } from "./gateway.service";
import { PeripheralController } from "./peripheral.controller";
import { PeripheralService } from "./peripheral.service";

@Module({
    imports: [
        TypeOrmModule.forFeature([Gateway, Peripheral])
    ],
    controllers: [GatewayController, PeripheralController],
    providers: [GatewayService, PeripheralService],
  })
export class GatewayModule {}