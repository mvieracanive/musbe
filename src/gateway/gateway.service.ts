import { HttpException, HttpStatus, Injectable, Options } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { GatewayDto } from '../dtos/gateway.dto';
import { Repository } from 'typeorm';
import { Gateway } from './entities/gateway.entity';

Injectable();
export class GatewayService {
  constructor(@InjectRepository(Gateway) private _repo: Repository<Gateway>) {}

  async create(dto: GatewayDto): Promise<GatewayDto> {
    try {
      await this._repo.insert(dto);
      return await this.getOne(dto.serial);
    } catch (ex) {
      if (ex.errno == 19)
        throw new HttpException(
          'Gateway serial must be a unique identifier',
          HttpStatus.CONFLICT,
        );
      throw new HttpException(
        'Internal Error. Please check the logs',
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  async getCollection(sort: string): Promise<GatewayDto[]> {
    let options = {};
    if (sort) options[sort] = 'ASC';
    
    const ret = this._repo.find({ order: options });

    return ret;
  }

  async getOne(id: string) {
    return this._repo.findOne(id);
  }
}
