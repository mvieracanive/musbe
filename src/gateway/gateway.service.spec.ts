import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken, TypeOrmModule } from '@nestjs/typeorm';
import { DatabaseModule } from '../database/database.module';
import { GatewayDto } from '../dtos/gateway.dto';
import { getConnection, Repository } from 'typeorm';
import { Gateway } from './entities/gateway.entity';
import { GatewayService } from './gateway.service';
import { ConnectionOptions } from 'tls';
import { Peripheral } from './entities/peripheral.entity';
import { throwError } from 'rxjs';
import { HttpException, HttpStatus } from '@nestjs/common';

type MockType<T> = {
  [P in keyof T]?: jest.Mock<{}>;
};
let data = {serial: 'aa', human_name:"aaname", ipv4: "100.456.7.7"};
let data2 = {serial: 'bb', human_name:"bbname", ipv4: "101.456.7.7"};
let dataArr = [
  data,
  data2
];

const repositoryMockFactory: () => MockType<Repository<any>> = jest.fn(() => ({
  findOne: jest.fn(id => {return {...data, serial: id}}),
  insert: jest.fn(entity => entity),
  find: jest.fn((sort) => dataArr),
}));

describe('GatewayService', () => {
  let service: GatewayService;
  let repositoryMock: any;
  

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
        imports: [
        ],
        controllers: [],
        providers: [
          { 
            provide: getRepositoryToken(Gateway), 
            useFactory: repositoryMockFactory
         },
          GatewayService],
    }).compile();

    service = app.get<GatewayService>(GatewayService);
    repositoryMock = app.get(getRepositoryToken(Gateway));
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('create', () => {    
    it('should return a Gateway', async () => {

      const res = await service.create(data);
      expect(res).toEqual(data);
      expect(repositoryMock.findOne).toHaveBeenCalledWith(data.serial);
      expect(repositoryMock.insert).toHaveBeenCalledWith(data);        
    });
  });

  describe('create', () => {    
    it('should throw an HttpStatus.CONFLICT when gateway serial already exists in DB', async () => {
      
      repositoryMock.insert.mockImplementation((entity) => Promise.reject({errno: 19}));
      try{
        await service.create(data);
      }
      catch(ex){      
        expect(ex).toBeInstanceOf(HttpException); 
        expect(ex).toMatchObject({status: HttpStatus.CONFLICT});
      }
      
      expect(repositoryMock.findOne).not.toHaveBeenCalled();
      expect(repositoryMock.insert).toHaveBeenCalledTimes(1);
      expect(repositoryMock.insert).toHaveBeenCalledWith(data);        
    });
  });

  describe('create', () => {    
    it('should throw an HttpStatus.INTERNAL_SERVER_ERROR when there is an error inserting in DB not clasified by the TRY/CATCH code', async () => {
      
      repositoryMock.insert.mockImplementation((entity) => Promise.reject('Unknow error'));
      try{
        await service.create(data);
      }
      catch(ex){      
        expect(ex).toBeInstanceOf(HttpException); 
        expect(ex).toMatchObject({status: HttpStatus.INTERNAL_SERVER_ERROR});
      }
      
      expect(repositoryMock.findOne).not.toHaveBeenCalled();
      expect(repositoryMock.insert).toHaveBeenCalledTimes(1);
      expect(repositoryMock.insert).toHaveBeenCalledWith(data);        
    });
  });

  describe('getCollection', () => {
    it('should return a list of GatewayDto', async () => {
      const sort: string = '';

      const res = await service.getCollection(sort);
      expect(res).toEqual(dataArr);
      expect(repositoryMock.find).toHaveBeenCalledWith({order: {}});        
    });
  });

  describe('getCollection', () => {
    it('should find gateways ordering ASC by serial', async () => {
      const sort: string = 'serial';

      const res = await service.getCollection(sort);
      expect(repositoryMock.find).toHaveBeenCalledWith({order: {serial: 'ASC'}});        
    });
  });

  describe('getCollection', () => {
    it('should find gateways ordering ASC by serial', async () => {
      const sort: string = 'human_name';

      const res = await service.getCollection(sort);
      expect(repositoryMock.find).toHaveBeenCalledWith({order: {human_name: 'ASC'}});        
    });
  });

  describe('getCollection', () => {
    it('should find gateways ordering ASC by serial', async () => {
      const sort: string = 'ipv4';

      const res = await service.getCollection(sort);
      expect(repositoryMock.find).toHaveBeenCalledWith({order: {ipv4: 'ASC'}});        
    });
  });
})