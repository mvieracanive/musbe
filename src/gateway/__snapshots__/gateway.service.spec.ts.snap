// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GatewayService getCollection should return a list of GatewayDto 1`] = `
Object {
  "human_name": Any<String>,
  "ipv4": Any<String>,
  "serial": Any<String>,
}
`;

exports[`GatewayService getCollection should return a list of GatewayDto 2`] = `
Object {
  "human_name": Any<String>,
  "ipv4": Any<String>,
  "serial": Any<String>,
}
`;
